[2D parallel] workaround for FSDP init issue

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103866

Closes https://github.com/pytorch/pytorch/issues/96491 and does so by relaxing FSDP's assumption that the entire input module must be on the same device. Now, FSDP can accept a module partially on CPU and GPU and just emits a warning.
Differential Revision: [D46847184](https://our.internmc.facebook.com/intern/diff/D46847184/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D46847184/)!

rohan-varma committed:
[2D parallel] workaround for FSDP init issue

Per title

Differential Revision: [D46847184](https://our.internmc.facebook.com/intern/diff/D46847184/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D46847184/)!

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -753,6 +753,12 @@ def _check_single_device_module(
+            "This could occur in cases where another parallelism algorithm has already"
+            "moved some parameters to a non-CPU device."
+        )
     if len(devices) > 1:
```
If `if len(devices) == 2 and torch.device("cpu") in devices == True`, I am not seeing why this `if len(devices) > 1:` branch would not also get taken, still leading to the `RuntimeError` getting raised.

rohan-varma replied:
yeah, we need to change this to an `else`

rohan-varma committed:
Update on "[2D parallel] workaround for FSDP init issue"


Closes https://github.com/pytorch/pytorch/issues/96491 and does so by relaxing FSDP's assumption that the entire input module must be on the same device. Now, FSDP can accept a module partially on CPU and GPU and just emits a warning.
Differential Revision: [D46847184](https://our.internmc.facebook.com/intern/diff/D46847184/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D46847184/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[2D parallel] workaround for FSDP init issue"


Closes https://github.com/pytorch/pytorch/issues/96491 and does so by relaxing FSDP's assumption that the entire input module must be on the same device. Now, FSDP can accept a module partially on CPU and GPU and just emits a warning.
Differential Revision: [D46847184](https://our.internmc.facebook.com/intern/diff/D46847184/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D46847184/)!

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -746,14 +746,34 @@ def _get_buffer_names(root_module: nn.Module) -> Set[str]:
+        if device_id is not None:
+            warnings.warn(
+                f"FSDP wrapping module with params on multiple devices: {devices}"
+                "This could occur in cases where another parallelism algorithm has already"
+                "moved some parameters to a non-CPU device."
```
I am not sure if we should give this warning.

What I am thinking is that one of the following cases holds:
- The user knows that he/she is using mixture of parallelisms, and then this warning is not too informative.
- The user does not know that he/she is using a mixture of parallelisms, and then this warning may be jarring.
- the user is not using a mixture of parallelisms but hits this branch somehow, and then this warning may be confusing.

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -746,14 +746,34 @@ def _get_buffer_names(root_module: nn.Module) -> Set[str]:
+            )
+        else:
+            raise RuntimeError(
+                f"To support partial CPU / GPU module, please pass in device_id argument."
```
nit: Maybe "To support a module with both CPU and GPU parameters"?

awgu commented on test/distributed/fsdp/test_fsdp_misc.py:
```diff
@@ -482,20 +482,53 @@ def test_module_device_mismatches_device_id(self):
+        """
+        Module on multiple GPUs wrapped in FSDP should raise an error.
+        """
+        torch.cuda.set_device(self.rank)
```
curious: Do we not already set this in the test setup? (I feel like we rely on it being set.)

fduwjj commented on test/distributed/tensor/parallel/test_2d_parallel.py:
```diff
@@ -50,22 +51,20 @@ def _distribute_and_fsdp_wrap_module(
-        )
-        module.net2 = FSDP(
-            module.net2, process_group=pg, use_orig_params=use_orig_params
-        )
-    return FSDP(module, process_group=pg, use_orig_params=use_orig_params)
+        module.net1 = fsdp_ctor(module.net1)
+        module.net2 = fsdp_ctor(module.net2)
+
+    return fsdp_ctor(module)
```
Thanks for cleaning this up.

fduwjj commented on test/distributed/tensor/parallel/test_2d_parallel.py:
```diff
@@ -50,22 +51,20 @@ def _distribute_and_fsdp_wrap_module(
     world_size = dist.get_world_size()
 
-    model = SimpleModel().cuda(rank)
+    model = SimpleModel()
```
Nice!

rohan-varma committed:
Update on "[2D parallel] workaround for FSDP init issue"


Closes https://github.com/pytorch/pytorch/issues/96491 and does so by relaxing FSDP's assumption that the entire input module must be on the same device. Now, FSDP can accept a module partially on CPU and GPU and just emits a warning.
Differential Revision: [D46847184](https://our.internmc.facebook.com/intern/diff/D46847184/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D46847184/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[2D parallel] workaround for FSDP init issue"


Closes https://github.com/pytorch/pytorch/issues/96491 and does so by relaxing FSDP's assumption that the entire input module must be on the same device. Now, FSDP can accept a module partially on CPU and GPU and just emits a warning.
Differential Revision: [D46847184](https://our.internmc.facebook.com/intern/diff/D46847184/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D46847184/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[2D parallel] workaround for FSDP init issue"


Closes https://github.com/pytorch/pytorch/issues/96491 and does so by relaxing FSDP's assumption that the entire input module must be on the same device. Now, FSDP can accept a module partially on CPU and GPU and just emits a warning.
Differential Revision: [D46847184](https://our.internmc.facebook.com/intern/diff/D46847184/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D46847184/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[2D parallel] workaround for FSDP init issue"


Closes https://github.com/pytorch/pytorch/issues/96491 and does so by relaxing FSDP's assumption that the entire input module must be on the same device. Now, FSDP can accept a module partially on CPU and GPU and just emits a warning.
Differential Revision: [D46847184](https://our.internmc.facebook.com/intern/diff/D46847184/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D46847184/)!

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI done"

rohan-varma committed:
Update on "[2D parallel] workaround for FSDP init issue"


Closes https://github.com/pytorch/pytorch/issues/96491 and does so by relaxing FSDP's assumption that the entire input module must be on the same device. Now, FSDP can accept a module partially on CPU and GPU and just emits a warning.
Differential Revision: [D46847184](https://our.internmc.facebook.com/intern/diff/D46847184/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D46847184/)!

[ghstack-poisoned]
